### PR TITLE
py-click: fix Python 3.6 support

### PR DESCRIPTION
**Pull request by adamjstewart**
CI is broken due to #40473 

@alalazo the test you added to ensure that Python 3.6 bootstrapping still works doesn't fail on the PR that breaks things, only on all subsequent PRs.

@tgamblin when we can stop supporting `spack style` bootstrapping on EOL Python versions? We can still support every other part of Spack for Python 3.6 if you want, but this is only going to get worse when we drop Python 3.7 this fall and 3.8 next year.

### CODE REVIEW

**Comment by haampie** [comment]
> but this is only going to get worse when we drop Python 3.7 this fall and 3.8 next year.

Who is "we" here? I think dropping 3.6 was pretty controversial. I don't see the point in removing older Python versions from Spack even if they're EOL, what matter is whether they're used or not, and they will continue to get used long after their EOL date. Dropping old Python versions makes it a pain to bootstrap Spack's own dependencies, as it is currently with clingo binaries for Python 3.6.

If you don't wanna maintain Python 3.7, why don't you set an upperbound like `depends_on("python@:3.7", when="@:current-latest-version")` or so?

**Comment by adamjstewart** [comment]
"we" as in the maintainers of `python` and `PythonPackage`, i.e., me. If anyone else wants to join, feel free. Many have joined in the past, but none have actually reviewed a single PR or helped fix these kind of issues.

I'm a postdoc now, I have limited time to maintain versions of Python released in 2016. If you want to use them to run Spack, that's fine, but if you want to use Spack to install them, you should really reconsider. They're unmaintained (by me or by PSF) and likely contain known security vulnerabilities. #33898 removed close to 8K lines of code I no longer have to maintain.

**Comment by haampie** [comment]
Is there that much overhead in supporting 3.7 / 3.8? As far as I can see, the python package itself hasn't changed significantly (except maybe 3.12 being configure script only), and for python packages, you have to verify the compat bounds for Python anyways.

I could see there would be more overhead in supporting both Python 2 and 3.

**Comment by adamjstewart** [comment]
Not in the Python package itself. It's more in the packages that depend on Python. It gives us an easy excuse to deprecate and remove all versions that don't support modern Python versions. We could of course do this on a package-by-package basis, but there's too many packages and only one me. Since we removed Python 3.6 and older, I don't think I've seen a single request to be able to install those versions. If no one is installing EOL versions of Python, I see no reason to support them.

**Comment by haampie** [comment]
Again, it's about Spack itself, it's (a) really poor user experience if Spack fails to bootstrap and (b) we as Spack developers have to jump through hoops to generate binaries for clingo etc, maybe @alalazo / me haven't been loud enough.

I don't really see how deleting lines `depends_on("python@3.7:", when="@x:")` reduces the maintenance burden of adding a new version `x+1`, it would be the same work, right? 🤔 aren't the rules for 3.7 compat mostly frozen forever in package.py files now?

**Review by haampie — APPROVED** [review]

**Comment by adamjstewart** [comment]
Completely agree, Spack needs to be able to bootstrap clingo, but clingo doesn't depend on click. If we can drop the `spack style` deps, I'll be happier.

The compromise we agreed on is that I can do whatever I want (within Python packaging) as long as bootstrap-dev-rhel8 continues to pass. So we just need to figure out why that's not correctly detecting issues before merging.

**Pull request merged**
